### DISCUSSION TITLE
fixed size buffer some problems?

### DISCUSSION THREAD

**Discussion by sgf** [dotnet/csharplang#5182] [Q&A] [answered]
```C#
    public interface IOption {
        public DHCPOptions Type { get; set; }
        public byte Length { get; set; }
        //public Span<byte> Value { get; }
    }
```

#### NO1. how to cteate a Span<byte> property from a fixed size buffer ?
```C#
public unsafe struct OptionClassId : IOption {
        public DHCPOptions Type { get; set; }
        public byte Length { get; set; }

        public fixed byte ClassId[128];
        //public Span<byte> Value => Unsafe.AsRef(ClassId);// Unsafe.Read<byte>(ClassId);

    }
```
#### NO2.how to use sizeof like C++ in C# get the size of fixed size buffer ?
like:
```C#
 //get size of the filed of ClassId
int classIdLenght= sizeof(ClassId);
```
#### NO3.how to create a Extension Method for fixed size buffer ?
like:
```C#
    public static unsafe Span<byte> AsSpan<T>(this  T* unmanagedT) where T : unmanaged {
            return new Span<byte>(unmanagedT, sizeof(T));
        }
ClassId.AsSpan()
```
#### NO4.how to create a custom struct as the fixed size buffer's base Type?
like:
```C#
    public struct OptionDomainNameServer : IOption {
        public DHCPOptions Type { get; set; }
        public byte Length { get; set; }
        public fixed IPv4 DomainNameServers[2];//?????
        public Span<byte> Value => throw new NotImplementedException();
    }

    public readonly struct IPv4 : IEquatable<IPv4> {
        private readonly uint value;
         //...
   }
```

**Comment by bernd5** [comment, marked as answer]
**NO1. how to cteate a Span property from a fixed size buffer ?**
You could write:

```csharp
using System;

struct Foo
{
    private const int BufLen = 128;
    private unsafe fixed byte Buffer[BufLen];
    public Span<byte> Data {
         get{
             unsafe {
                 fixed(Byte* ptr = &Buffer[0]){
                     return new Span<byte>(ptr, BufLen);
                 }    
             }
         }
    }
}
```

**NO2.how to use sizeof like C++ in C# get the size of fixed size buffer ?**
There is no `sizeof` support. The type of an fixed buffer access is just a pointer.

**NO3.how to create a Extension Method for fixed size buffer?**
You can't create extension method for pointers -> you can't create extensionmethods for fixed size buffers.

**NO4.how to create a custom struct as the fixed size buffer's base Type?**
I don't understand that question. But the base type of structs is always `System.ValueType`. The possible types in fixed size buffers is limited: https://github.com/dotnet/csharplang/blob/3c8559f186d4c5df5d1299b0eaa4e139ae130ab6/spec/unsafe-code.md#fixed-size-buffer-declarations. But there is a proposal to relax this: https://github.com/dotnet/csharplang/blob/main/proposals/fixed-sized-buffers.md.

**Reply by yaakov-h** [reply]
Isn't it dangerously unsafe to return a Span that outlives the `fixed` block of its pointer?

**Reply by sgf** [reply]
Thank you very much, patience to answer.

**Reply by jnm2** [reply]
> Isn't it dangerously unsafe to return a Span that outlives the fixed block of its pointer?

Yes. The GC may move the thing containing the struct around on the heap at any time and the pointer will be pointing at something random. It's not possible to safely use the pointer when pinning ends, which in this case is at the end of the `fixed` block.

See the approach taken by CsWin32 (a new official vehicle for consuming Windows platform APIs): https://github.com/microsoft/CsWin32/issues/117#issuecomment-783794356. Tanner and Jared comment in that thread.
